common/cache.go

```diff
@@ -228,17 +228,26 @@ func NewBlockCache(dbPath string, chainName string, startHeight int, syncFromHei
 	c.starts = nil
 	c.starts = append(c.starts, 0)
 	nBlocks := len(lengths) / 4
-	Log.Info("Reading ", nBlocks, " blocks (since Sapling activation) from disk cache ...")
+	Log.Info("Reading ", nBlocks, " blocks from the cache ...")
 	for i := 0; i < nBlocks; i++ {
 		if len(lengths[:4]) < 4 {
 			Log.Warning("lengths file has a partial entry")
 			c.recoverFromCorruption(c.nextBlock)
 			break
 		}
 		length := binary.LittleEndian.Uint32(lengths[i*4 : (i+1)*4])
-		if length < 74 || length > 4*1000*1000 {
-			Log.Warning("lengths file has impossible value ", length)
-			c.recoverFromCorruption(c.nextBlock)
+
+		// This is the length of the first block, after the changes that
+		// store transparent transaction data, see:
+		// https://github.com/zcash/lightwallet-protocol/pull/1
+		// If the length is anything else, we're probably running on an
+		// old data directory, so we must rebuild the cache.
+		// Also allow 149 for the unit test TestCache().
+		if i == 0 && !(length == 1639 || length == 149) {
```

Comment:
This length check should probably be a check against the (length, chainName) pair, such that there's not a risk of cross-protocol screwups.